lib: posix: Add Posix Style File System API support

ramakrishnapallala opened the pull request:
Add support for POSIX style file system API's
Add a test to verify the Posix FS API's

Commit:
lib: posix: Add Posix Style File System API support

Add IEEE 1003.1 Posix Style file system API support.
These API's will internally use corresponding Zephyr
File System API's.

Signed-off-by: Ramakrishna Pallala <ramakrishna.pallala@intel.com>

pfalcon commented:
> Note: At present we can not use these API's with NewLibC
enabled as it is also defining the open, read, write, lseek and
close functions in libc-hooks.c file.

Yes, so we can't really seriously talk about POSIX support with a good libc support (libc is part of POSIX). So, almost certainly, we would need implement filesystem functions on newlib level, using libc-hooks.c and friends.

Were there other requirements? Where are they, I don't see any links to github issues containing them in the commit messages.

Thanks.

pfalcon commented:
> include/posix/fs.h

POSIX API compliance also includes using standard header file names, e.g. read() is defined in `unistd.h`.

pfalcon commented on include/posix/fs.h:
```diff
@@ -0,0 +1,55 @@
+extern int open(const char *name, int flags);
+extern int close(int file);
+extern int write(int file, char *buffer, unsigned int count);
+extern int read(int file, char *buffer, unsigned int count);
```
POSIX defines return value to be of type ssize_t. Similarly with params and other functions.

ramakrishnapallala commented:
@pfalcon we have taken up the file system API support as an incremental feature to ongoing Posix/Pthread API support effort.

I think there is no Github issue for this. Here is the related issue:
https://github.com/zephyrproject-rtos/zephyr/issues/1792

nashif commented:
without support for newlib, this is not going anywhere.

ramakrishnapallala commented:
@pfalcon libc-hooks.c is implementing open(), close() and friends. So if we want to enabling FS API's then do we have to modify those functions with implementation in lib/posix/fs.c (in this PR)?

Do we need the same FS API support in case minimal libc?

pfalcon commented:
> How I got here is because BSD Sockets API also defines close() function, so would be incompatible with this API.

I captured this and other requirements for supporting sockets as file descriptors here: https://github.com/zephyrproject-rtos/zephyr/issues/7405

ramakrishnapallala commented:
@pfalcon thanks for your inputs. I will get back to you once I dig more into the details.

nashif commented:
This needs to work with any libc, not only newlib. Someone might chose to use their own libc and not newlib. In any case, the usage of those calls with bsd sockets and other subsystems should be looked at, but we should not block this IMO. We need a proposal for supporting file descriptors and owner for this, I do not feel however we should wait for this to be done to get file system support in.

Commit:
tests: posix: Add test for Posix File System API's

Added a test to verify Posix File System API's.

Signed-off-by: Ramakrishna Pallala <ramakrishna.pallala@intel.com>

ramakrishnapallala commented:
@pfalcon I have updated PR and disabled libc-hooks if posix FS is enabled. Can you have a look and let me know it that fine. I don't have the background on the install_hooks code, please explain the purpose of those hooks and which app/tests are using it.

lpereira commented on lib/posix/fs.c:
```diff
@@ -0,0 +1,267 @@
+		return -EBADF;
+	}
+
+	rc = fs_closedir(dirp);
```
dirp should be validated as well.  If you end up using the suggestion of storing the used bit in a bitmap rather than an entry in the struct, you can use the PART_OF_ARRAY macro to check for validity

ramakrishnapallala replied:
ok

lpereira commented on lib/posix/fs.c:
```diff
@@ -0,0 +1,267 @@
+
+	rc = fs_readdir(dirp, &fdirent);
+	if (rc < 0) {
+		return NULL;
```
Shouldn't errno be converted from rc?

lpereira commented on lib/posix/fs.c:
```diff
@@ -0,0 +1,267 @@
+	((struct p_fs_dir *)dirp)->used = false;
+	k_mutex_unlock(&dirp_mutex);
+
+	return rc < 0 ? -1 : 0;
```
errno should be set

lpereira commented on lib/posix/fs.c:
```diff
@@ -0,0 +1,267 @@
+	int rc;
+
+	rc = fs_rename(old, new);
+	return rc < 0 ? -1 : 0;
```
Set errno from rc.

lpereira commented on lib/posix/fs.c:
```diff
@@ -0,0 +1,267 @@
+	int rc;
+
+	rc = fs_unlink(path);
+	return rc < 0 ? -1 : 0;
```
Set errno from rc.

lpereira commented on lib/posix/fs.c:
```diff
@@ -0,0 +1,267 @@
+
+	rc = fs_statvfs(path, &stat);
+
+	return rc;
```
Set errno from rc.

lpereira commented on lib/posix/fs.c:
```diff
@@ -0,0 +1,267 @@
+	ARG_UNUSED(mode);
+
+	rc = fs_mkdir(path);
+	return rc < 0 ? -1 : 0;
```
Set errno from rc.

lpereira commented on lib/posix/fs.c:
```diff
@@ -0,0 +1,267 @@
+	bool used;
+};
+
+static struct p_fs_dir dir_array[CONFIG_MAX_OPEN_FILES];
```
You might want to consider using an union to store both files and directories:

```C
union file_descriptor {
    struct fs_file_t file;
    struct fs_dir_t dir;
};
```

The reason is that you'll want to support `dirfd()` in the future, and you'll need to unify both directory and file descriptors.  This will also simplify some things (e.g. only one mutex), but you'll need another array to store flags/type:

```C
struct file_descriptor_flags {
    u8_t used : 1;
    u8_t is_directory : 1;
    u8_t refs : 6;
};
STATIC_ASSERT(CONFIG_MAX_OPEN_FILES < 1<<6);
static struct file_descriptor_flags fd_flags[CONFIG_MAX_OPEN_FILES];
```

It's also possible to write a function such as: `int find_free_fd(void)` and use it for both open() and opendir().

Another thing to notice is that there's no distinction between a file descriptor (per-process) and file description (per system).  This is fine for an embedded RTOS such as Zephyr, but we need to take into consideration a few other things.  For instance, removing files while they're open is something that should be supported, but no operation should be sent to the file system while they're open; the system must keep track of this and enforce refcounts.

ramakrishnapallala replied:
I had treated open files and directories separately but this would combine them and have one open fd list which makes sense from unix point of view. Will fix it accordingly.

ramakrishnapallala commented on lib/posix/fs.c:
```diff
@@ -0,0 +1,267 @@
+	k_mutex_unlock(&filep_mutex);
+
+	if (fd >= CONFIG_MAX_OPEN_FILES) {
+		return -1;
```
do you mean set _current->errno_var under CONFIG_ERRNO?

lpereira replied:
Yes, but just do `errno = SOME_ERROR_CODE`.  `errno` is a macro that expands to the correct thing.

ramakrishnapallala commented on lib/posix/fs.c:
```diff
@@ -0,0 +1,267 @@
+{
+	int rc;
+
+	if (fd >= CONFIG_MAX_OPEN_FILES) {
```
ok

ramakrishnapallala commented on lib/posix/fs.c:
```diff
@@ -0,0 +1,267 @@
+ssize_t write(int fd, char *buffer, unsigned int count)
+{
+	int rc;
+
```
Ok

ramakrishnapallala commented on lib/posix/fs.c:
```diff
@@ -0,0 +1,267 @@
+ */
+ssize_t read(int fd, char *buffer, unsigned int count)
+{
+	int rc;
```
Ok will fix it.

andyross replied:
Meta-pedantry: ssize_t is larger than or equal to int on all platforms, and both are guaranteed to be signed.  This code is safe as it is, though style and clarity would argue for @lpereira's nit.

ramakrishnapallala commented on lib/posix/fs.c:
```diff
@@ -0,0 +1,267 @@
+static struct p_fs_file file_array[CONFIG_MAX_OPEN_FILES];
+K_MUTEX_DEFINE(filep_mutex);
+
+static struct fs_dirent fdirent;
```
No.

lpereira commented on lib/posix/fs.c:
```diff
@@ -0,0 +1,267 @@
+ *
+ * See IEEE 1003.1
+ */
+struct dirent *readdir(DIR *dirp)
```
Might be a good idea to support readdir_r() as well, and implement readdir() in terms of it:

```C
struct dirent *readdir_r(DIR *dirp, struct dirent *entry, struct dirent **result)
{
}

struct dirent *readdir(DIR *dirp)
{
       struct dirent *ent = &pdirent;
       int rc = readdir_r(dirp, ent, &ent);

       if (!rc) {
             return ent;
       }

       errno = -rc;
       return 0;
}
```

ramakrishnapallala replied:
I will defer this as an enhancement.

andyross commented on include/posix/unistd.h:
```diff
@@ -11,11 +11,43 @@ extern "C" {
+
+struct dirent {
+	unsigned int d_ino;
+	char *d_name;
```
Almost always d_name is a fixed-size array inside struct dirent.  This isn't incorrect per the spec AFAIK, but it's a little weird and it's not impossible that it might break code that does something weird like trying to write over &d_name instead of &d_name[0]

andyross commented on lib/posix/fs.c:
```diff
@@ -0,0 +1,267 @@
+K_MUTEX_DEFINE(dirp_mutex);
+
+static struct p_fs_file file_array[CONFIG_MAX_OPEN_FILES];
+K_MUTEX_DEFINE(filep_mutex);
```
Why a mutex?  Unless you really need priority inheritance (and AFAICT you don't), you don't want this.  Use either a semaphore (if you want blocking - again AFAICT you don't) or an irq_lock() (if you can get away with just a critical section).  Or if you want to be fancy use a spinlock, which becomes a non-recursive irq_lock() in uniprocessor mode and is more futureproof.

andyross commented on lib/posix/fs.c:
```diff
@@ -0,0 +1,267 @@
+	for (fd = 0; fd < CONFIG_MAX_OPEN_FILES; fd++) {
+		if (file_array[fd].used == false) {
+			ptr = &file_array[fd].file;
+			memset(ptr, 0, sizeof(struct fs_file_t));
```
Performance nit: put the memset and all initialization outside the lock.  The locking should be to find an open slot and reserve it.  Then you can touch the file record without synchronization because nothing else will use it.

andyross commented on lib/posix/fs.c:
```diff
@@ -0,0 +1,267 @@
+	if (rc < 0) {
+		k_mutex_lock(&filep_mutex, K_FOREVER);
+		file_array[fd].used = false;
+		k_mutex_unlock(&filep_mutex);
```
If all you're doing is marking the record as free using an atomic type, you don't need synchronization here.  There's no intermediate state that anyone can get confused by.  That's true as long as "used" is a byte -- if you made it a bitmask you'd need some kind of synchronization of course.

andyross commented on lib/posix/fs.c:
```diff
@@ -0,0 +1,267 @@
+	/* Free file ptr memory */
+	k_mutex_lock(&filep_mutex, K_FOREVER);
+	file_array[fd].used = false;
+	k_mutex_unlock(&filep_mutex);
```
Same deal.  No need for locking if you're just changing one value in a critical section.

andyross commented on lib/posix/fs.c:
```diff
@@ -0,0 +1,267 @@
+	struct fs_dir_t *ptr;
+
+	k_mutex_lock(&dirp_mutex, K_FOREVER);
+	/* Allocate memory for dir ptr */
```
This comment is sort of a lie.  "Allocate memory" implies a heap somewhere, this is just finding an open slot.

andyross commented on lib/posix/fs.c:
```diff
@@ -0,0 +1,267 @@
+			ptr = &dir_array[fd].dir;
+			memset(ptr, 0, sizeof(struct fs_dir_t));
+			dir_array[fd].used = true;
+			break;
```
This loop is identical to the "find an open slot" code in open() above.  Please unify.

It's also susceptible to the same criticism about locking I had there, which I won't type twice because it should only appear once in the file anyway. :)

andyross commented on lib/posix/fs.c:
```diff
@@ -0,0 +1,267 @@
+	if (rc < 0) {
+		k_mutex_lock(&dirp_mutex, K_FOREVER);
+		dir_array[fd].used = false;
+		k_mutex_unlock(&dirp_mutex);
```
Third time: no locking please.

andyross commented on lib/posix/fs.c:
```diff
@@ -0,0 +1,267 @@
+	/* Free file ptr memory */
+	k_mutex_lock(&dirp_mutex, K_FOREVER);
+	((struct p_fs_dir *)dirp)->used = false;
+	k_mutex_unlock(&dirp_mutex);
```
Fourth time, no locking.  Also, why the cast?  Seems like DIR and struct ps_fs_dir* should be a typedef already, no?

andyross commented on tests/posix/fs/src/test_fs_file.c:
```diff
@@ -0,0 +1,162 @@
+		return brw;
+	}
+
+	read_buff[brw] = 0;
```
You need to test brw against sizeof(read_buff) or you'll overflow here.

andyross commented on tests/posix/fs/src/test_fs_file.c:
```diff
@@ -0,0 +1,162 @@
+		return res;
+	}
+
+	TC_PRINT("Closed file %s\n", TEST_FILE);
```
Did you?  Seems like this test needs an extra step to induce and detect an error on some other I/O on that file descriptor.  Right now it just tests that close() returns success, which isn't the same thing, really.

andyross reviewed (changes requested):
Mostly just nitpicking, but using a priority-inheriting mutex (instead of irq_lock()) to protect a trivial critical section around an bunch of "allocated" flags rises to the level of a fatal  design flaw in my book.

Also: I note there's no synchronization of the I/O calls themselves. Multiple threads banging on a single file descriptor will race against themselves into the underlying filesystem code (which IIRC is not threadsafe either).  Is that OK?  Maybe it is... should be documented though.

ramakrishnapallala commented:
@andyross @lpereira I have addressed your comments and updated the PR. please have a look at it.

@aescolar-ot I am getting compilation errors if I build for native_posix. I have added the POSIX API's which I have added to this PR in posix_cheats.h file but still getting compilation issue. For the errors, you look at the shippable logs.

andyross commented on lib/posix/fs.c:
```diff
@@ -0,0 +1,399 @@
+#ifdef CONFIG_ERRNO
+static inline void set_pthread_errno(int err)
+{
+	unsigned int key = irq_lock();
```
Don't see why this locking is needed.  Surely pthread_self() is threadsafe, and errno_var is an atomic type.

lpereira replied:
Just set `errno`.  `errno` is not a variable, but rather a macro that does the right thing.

andyross reviewed (approved):
One more nitpick.  Can't see anything else that should stop this

lpereira commented on include/posix/unistd.h:
```diff
@@ -11,11 +11,43 @@ extern "C" {
+
+struct dirent {
+	unsigned int d_ino;
+	char d_name[MAX_FILE_NAME + 1];
```
Might want to define PATH_MAX instead in limits.h, that is the standard value.  This minor, though, so please create an issue to track this.

lpereira commented on lib/posix/fs.c:
```diff
@@ -0,0 +1,399 @@
+#define set_pthread_errno(x) do { } while (0)
+#endif
+
+static int posix_fs_alloc_fd(void **ptr, bool is_dir)
```
For type safety and clarity, it's fine to use `union **file_desc` instead of `void **ptr`...

lpereira commented on lib/posix/fs.c:
```diff
@@ -0,0 +1,399 @@
+
+	for (fd = 0; fd < CONFIG_MAX_OPEN_FILES; fd++) {
+		if (desc_array[fd].used == false) {
+			if (is_dir) {
```
...and remove the branch here, setting only `*file_desc` regardless of the value of `is_dir`.  Whoever calls `posix_fs_alloc_fd()` will know if it should be used as a directory or as a file.

ramakrishnapallala replied:
ok

lpereira commented on lib/posix/fs.c:
```diff
@@ -0,0 +1,399 @@
+	irq_unlock(key);
+}
+
+static void posix_fs_free_ptr(void *ptr)
```
This is minor, but it's the same thing as `posix_fs_free_fd()`.  Can you write one in function of the other?  `posix_fs_free_fd()` can call `posix_fs_free_ptr()`.  They're both very small functions and it's OK to have copies, but since locking is involved, it's a good thing to not scatter the responsibility of using them.

ramakrishnapallala replied:
will fix it.

lpereira commented on lib/posix/fs.c:
```diff
@@ -0,0 +1,399 @@
+	irq_unlock(key);
+}
+
+static int posix_fs_get_ptr(int fd, void **ptr, bool is_dir)
```
Same thing about having a `union **` instead of a `void **`.

ramakrishnapallala replied:
OK

lpereira commented on lib/posix/fs.c:
```diff
@@ -0,0 +1,399 @@
+
+	if ((desc_array[fd].used == true) &&
+		(desc_array[fd].is_dir == is_dir)) {
+		if (is_dir) {
```
Same thing about this branch not being needed.

ramakrishnapallala replied:
ok

lpereira commented on lib/posix/fs.c:
```diff
@@ -0,0 +1,399 @@
+{
+	int rc;
+
+	rc = fs_unlink(path);
```
If the file is open, it can't be removed from the filesystem.  File descriptor allocation must have a reference count, and unlink() should only set a flag, that's checked during the time a file is closed, and only then it's passed to the filesystem.  Unless, of course, there's just one reference, in which case you can remove the file at that moment.

ramakrishnapallala replied:
As of now, this is not being handled and the assumption is application will do the due diligence of proper file open and close sequence.

lpereira commented on lib/posix/fs.c:
```diff
@@ -0,0 +1,399 @@
+ *
+ * See IEEE 1003.1
+ */
+int close(int fd)
```
Handles are not released when threads exit.  Zephyr does not have the notion of processes, so this is most likely fine as it's the expected behavior.  What does the POSIX spec say on the subject, especially when targeting embedded devices without processes?

lpereira reviewed (changes requested):
Patch looks a lot cleaner, thanks! There are some cleanups and enhancements I'd like to see, though, in addition to @andyross's comment.

ramakrishnapallala commented:
@lpereira I have addressed the comments including Andy's and updated the PR. Please review it.

lpereira commented on include/posix/unistd.h:
```diff
@@ -11,11 +11,46 @@ extern "C" {
+#ifdef CONFIG_POSIX_FS
+#include <fs.h>
+
+#undef PATH_MAX
```
`PATH_MAX` should be defined in `limits.h`, not `unistd.h`.  Redefining it here will mean that, if some code uses `limits.h` and not `unistd.h`, and it interacts with some code that uses `unistd.h`, they'll most likely be using different values.

`PATH_MAX` can be redefined only if it's being redefined to, say, `CONFIG_PATH_MAX`, so that we can make sure that it's the same value everywhere in the system, while making it configurable to boot (which is a good thing: a `struct dirent` might be allocated on the stack, and 256 bytes is too much for the tiny devices; in fact, that's as much kernel stack ARM devices have to handle syscalls).

ramakrishnapallala replied:
In limits.h (usr/incude/limits.h) is set as 4096 bytes so I have create another header file like "sys/type.h" for limits.h. For now, we keep in unistd.h. I will do the needed changes along with the next set of requirements.

lpereira replied:
This is a security vulnerability waiting to happen.

lpereira reviewed (approved):
LGTM with a comment
